### DISCUSSION TITLE
Diesel migration does not accept IDENTITY as valid code

### DISCUSSION THREAD

**Discussion by Davoodeh** [diesel-rs/diesel#3415] [Q&A] [answered]
### Versions

- **Rust:** `rustc 1.65.0 (897e37553 2022-11-02)`
- **Diesel:** `diesel 2.0.1`
- **Database:** `sqlite`
- **Operating System:** i use arch btw

## Problem Description

Migration of valid ANSI SQL fails to run. The code fails on `GENERATED ALWAYS AS IDENTITY` and `GENERATED BY DEFAULT AS IDENTITY` which are THE standard ways to make keys as far as I know.

Do Diesel even aim to be standard? If not why? It's an ORM after all, how is one to maintain multiple codes for different backends (sqlite and postgres for example) in Diesel if only supports one `up.sql` per migration and does not even support standard SQL?

### What are you trying to accomplish?

Trying to run `diesel migration run` on a simple minimal example of migrations as below:

```sql
-- up.sql
CREATE TABLE words (
    id INTEGER PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
    word VARCHAR NOT NULL
);
```

### What is the expected output?

`Running migration ...` and exit 0 (successful)

### What is the actual output?

`Failed to run ... with: near "IDENTITY": syntax error` with exit 1

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues)  and the [disussion forum](https://github.com/diesel-rs/diesel/discussions)  for similar possible closed issues.

- [x] This issue can be reproduced on Rust's stable channel.
- [x] This issue can be reproduced without requiring a third party crate

**Comment by weiznich** [comment, marked as answer]
Migrations are passed literally to the underlying database system. This means diesel itself does not do anything with them. If you get an error that something is unsupported that means the underlying database system does not support this specific syntax. If you feel that's wrong, please consider opening an issue in their issue tracker, as diesel cannot change that.

I've also converted your issue to a discussion as:
* Our issue tracker is not a support forum
* This is clearly not a bug in diesel, but just some invalid SQL for the given database system. Checkout the SQLite documentation for details what kind of create table statements they support.

**Reply by Davoodeh** [reply]
Thank you for your detailed explanation.
As I have not studied Diesel code, I did not know that it does not edit the code in any ways whatsoever.

Some question remain to be answered, what feature does Diesel has or is planned to unify the SQLs? Don't you think this is a logical feature to ask from an ORM like Diesel since that is basically what an ORM do, making database related tasks unified under one syntax, rule-book and code base?

Unfortunately, I am not skilled enough to contribute this feature, but assuming somebody else is, can a PR like this be merged or you think it is not in the scope of this project?

Or maybe I'm jumping too early on the problem again (is it obvious I didn't read the docs?), does Diesel already have such a feature? Idk something like multiple file support (e.g. `up-sqlite.sql`) or an automatic code conversion? If yes, I really think this should be the default approach taken in the "Getting Started". 

Excuse me if I'm not to ask these questions in the continuation of this topic. I'll open another discussion for this if necessary.

**Reply by weiznich** [reply]
> Some question remain to be answered, what feature does Diesel has or is planned to unify the SQLs?

Diesel does provide limited abilities to write queries in a common SQL subset using the provided DSL. Please check the documentation to see what parts of SQL are covered by the provided SQL. To be clear here: Not everything supported in the DSL is supported by all database backends. We check that at compile time and reject code that's not supported for certain backends. Also it's not a goal of diesel to provide all functionality for all backends, but rather allow users to write highly backend specific queries to unlock the good performance. As for migrations: They are not covered by the DSL and it's an explicit design decision to support only manually written SQL migrations in diesel itself. This allows you to control your schema in whatever detail is required for your use case. Third party crates might provide their own migration integration into diesels migration system. Such a third party crate could implement a database agnostic version, but I'm not aware of any. 

In addition: Please do not try to transfer your knowledge from ORM's in other languages to diesel. Diesel was explicitly designed differently. You will likely run in trouble quite soon if you just try to use it like a classical ORM.

> Or maybe I'm jumping too early on the problem again (is it obvious I didn't read the docs?), does Diesel already have such a feature? Idk something like multiple file support (e.g. up-sqlite.sql) or an automatic code conversion? If yes, I really think this should be the default approach taken in the "Getting Started".

Just use different migration folders for different backends? And no, that's not a topic for "Getting Started" in my opinion, as most users do not really care about supporting multiple database backends at once. Again: It's an explicit design decision to use hand written SQL for the migrations. So something like "automatic code conversion" is something that is directly opposing some fundamental vision of how diesel should work in our eyes. (Again: Third party crates can provide whatever they want and that would be the way to go for someone who wants to implement such a feature)

**Reply by Davoodeh** [reply]
Thanks again for your awesome work and explanation.

That is what I wanted to hear and answers all my questions. About the vision, philosophy and how they will probably be in the future.
